week-1-ts-ecosystem/w01e04-light-me-up/src/solution2.ts

```diff
@@ -0,0 +1,59 @@
+/**
+ * W zasadzie tutaj możesz modyfikować wszystko poza wartościami:
+ *
+ * - 100 w przypadku PowerSource.
+ * - 20 w przypadku LightBulb.
+ *
+ * Pamiętaj `PowerSource` ma nie wiedzieć — kto z niego korzysta!
+ * Dodatkowo — rzucać błąd, jeśli zapas mocy się wyczerpie.
+ * */
+
+class PowerSource {
+    private energySupply = 100
+
+    consume(energy:number) {
+        this.notEnoughEnergyException()
+        this.energySupply -= energy
+    }
+
+    notEnoughEnergyException(){
+        if(this.energySupply === 0){
+            throw new Error('There is not enough energy to light your device.')
+        }
+    }
+}
+
+class LightBulb {
+    protected readonly powerConsumption = 20
+    constructor(powerSource: PowerSource) {}
+
+    light() {
+        powerSource.consume(this.powerConsumption);
+    }
+
+    delay(){
+        return new Promise((res,rej)=>{
+            try {
```

Comment:
coś nie jestem przekonany że ten `try...catch` zadziała poprawnie.

Reply:
Sprawdziłem - nie działa poprawnie. Pomimo tego, że rzucasz wyjątek to go nie łapiesz.
`try...catch` jest w złym miejscu. powinien otaczać `this.light()`

![image](https://user-images.githubusercontent.com/25398949/193394356-843e87bc-a3e5-48e9-832f-9459579faf53.png)

To powinno być złożone przy wywołaniu `this.light()`

Reply:
Teraz zwarany jest exit kod `1` co potwierdza i też sam fakt że w ten sposób się to pokazuje - że nie obługujesz tego wyjątku.
Mimo że masz kod w linii `:58`

```javascript
lightBulb1.lightNSec(6).catch(e=>console.log(e))
```

ten `.catch` nigdy się nie wywołuje.